### ISSUE TITLE
[D417] Documenting Functions that `Unpack`

### ISSUE THREAD

**Issue by JP-Ellis** [astral-sh/ruff#8774]
## Summary

[PEP 692](https://peps.python.org/pep-0692/) introduced a way define the keyword arguments are runtime using `Unpack` on a `TypedDict`. This allows for the `TypedDict`'s attributes to be used directly as keyword arguments, and the actual keyword argument shouldn't be used.

This causes issues when Ruff tries to check that the arguments are documented in a docstring.

## Example

The PEP document gives the following example which suffices for this example:

```python
from typing import TypedDict, Unpack

class Movie(TypedDict):
    name: str
    year: int

def foo(**kwargs: Unpack[Movie]) -> None:
    """
    Foo a movie.

    Args:
        name: The name of the movie.
        year: The year the movie was released.
    """
    assert kwargs["year"] > 0

movie: Movie = {"name": "Life of Brian", "year": 1979}

foo(movie)                                  # Invalid
foo(kwargs=movie)                           # Invalid
foo(**movie)                                # OK!
foo(name="The Meaning of Life", year=1983)  # OK!
```

In effect, the use of `Unpack[Movie]` is meant to replace the `kwargs` with the relevant keyword arguments as inferred from the `Movie` `TypedDict` and as a result:

```python
def foo(**kwargs: Unpack[Movie]) -> None: ...
```

becomes essentially equivalent to:

```python
def foo(*, name: str, year: int) -> None: ...
```

The above example (using Google style) will raise the following error:

```log
ruff-unpack-example.py:7:5: D417 Missing argument description in the docstring for `foo`: `**kwargs`
```

### Ideal World

The most ideal solution would be for Ruff to parse the `TypedDict` and adjust the keyword arguments accordingly. I suspect this is beyond the scope of Ruff's current capabilities.

### Compromise

If Ruff sees specifically the pattern `**{arg}: Unpack[{class}]`, then:

- Ruff should _not_ require that `arg` be documented.
- Better yet, Ruff should perform a check in case `arg` is documented, and raise a warning as the documentation is misleading.

**Comment by zanieb** [comment]
Thanks for this well written issue!

The example seems a little weird to me, wouldn't the parameters be documented in `Movie` rather than each function that uses it?

**Comment by JP-Ellis** [comment]
The specific issue raised is in relation to [`D417` Undocumented Param](https://docs.astral.sh/ruff/rules/undocumented-param/) which targets specifically the documenting (or lack thereof) of function arguments, and therefore is independent of `Movie` being documented or not.

Your suggestion of having the parameters documented in the `Movie` class rather than each function is exactly the 'Ideal World' solution I mention above. The problem with this is that it would require some quite deep understanding of Python code from Ruff; and I am not sure that Ruff does/can do this. What would happen if the `Movie` class is defined in a separate Python module or library? Should Ruff parse all imports? This is why I mentioned it as an 'Ideal World' solution.

**Comment by zanieb** [comment]
Is there even an editor which supports reading docstrings from the unpacked class? It looks like PyCharm has no support for `Unpack` and VSCode will only document the parameters if you duplicate them in the function.

Given that VSCode requires the example format we should avoid raising an error if it's documented that way. We should probably not require it though, allowing `**kwargs: ...` still.

**Comment by JP-Ellis** [comment]
I don't quite understand your point here, but I'll try and respond and feel free to correct any misunderstanding.

> Is there even an editor which supports reading docstrings from the unpacked class?

I do not know what you mean specifically. So I will answer a couple of interpretations:

- Are you asking whether editors will auto-suggest the various keyword arguments based on parsing the `TypedDict`? If so, VSCode supports this as explained in [this completed feature request](https://github.com/microsoft/pylance-release/discussions/3046)

- Are you asking whether editors will read the `TypedDict`'s docstrings and update the function's docstring? I'm not aware of any which do this, nor do I know how this would be achieved. In any case, the function's docstring should probably indicate how the function should be used (as opposed to how it is literally written).

In case there's any confusion (and I've added the below to the original post), from a typing and documentation perspective, the following should be equivalent:

```python
def foo(**kwargs: Unpack[Movie]) -> None: ...
def bar(*, name: str, year: int) -> None: ...
```

> It looks like PyCharm has no support for `Unpack`

This would indicated that PyCharm does not support [PEP 692](https://peps.python.org/pep-0692/). There is in fact [a tracked issue](https://youtrack.jetbrains.com/issue/PY-55044) requesting for this to be implemented properly.

> VSCode will only document the parameters if you duplicate them in the function.

This would indicate that VSCode is incorrect here, as I don't think this is compliant with PEP 692. I guess this might also depend on which extension you are using, since, as mentioned above, Pylance does support parsing `TypedDict` and updating the suggested keyword arguments based on this. 

> Given that VSCode requires the example format we should avoid raising an error if it's documented that way.

I quite disagree with this. I think we should base Ruff's behaviour on how Python is intended to work, as per the PEPs.

**Comment by zanieb** [comment]
Hm I think we did indeed misunderstand each other so I'll try again.

I was attempting to agree with your original suggestion. Basically, we should allow `kwargs` to be missing if `Unpack` is used. When we can find the attributes of the inner type, we should enforce that they are each present in the docstring — we cannot do so if the definition is in another file at this time. I'm unsure if this should be required by default, since in an "ideal world" the members would be documented on the inner type and displayed by the IDE. If someone prefers to use `kwargs` in their documentation, we should probably allow that instead of forcing them to copy the definition repeatedly.

**Comment by JP-Ellis** [comment]
Yeah, all good then, completely agree :)

I think for now the time being, it would be sufficient to adjust `D417` to ignore any argument which correspond to an `Unpack[...]` argument (whether it is called `kwargs` or not).

**Comment by phongddo** [comment]
Hey @zanieb 👋 I've run into the same issue in my current project. Even though this issue's two years old, I can take it on if you don’t mind 🙏

**Comment by zanieb** [comment]
@ntBre is a better person now :)

**Comment by ntBre** [comment]
Sounds good to me! I think it makes sense to suppress `D417` for parameters with `Unpack` annotations.

**Comment by phongddo** [comment]
Hey @ntBre 👋 thanks for taking a look! I've opened https://github.com/astral-sh/ruff/pull/21816 to suppress the rule for params with `Unpack` annotation. Please review it when you have a moment. Thanks! 🙏
